File: reference/whenwords/src/human-date.ts
/**
 * Returns a contextual date string based on how far the timestamp is from
 * the reference:
 *
 *   Same day:           "Today"
 *   ±1 day:             "Yesterday" / "Tomorrow"
 *   2–6 days past:      "Last Wednesday"
 *   2–6 days future:    "This Wednesday"
 *   Same year, >6 days: "March 5"
 *   Different year:     "March 5, 2023"
 *
 * All comparisons use UTC dates. Reference timestamp is always explicit.
 *
 * @node human-date
 * @contract human-date.test.ts
 * @hint pattern: Pure function, no system clock. All date math is UTC.
 */
export function humanDate(timestamp: number, reference: number): string {
  const DAY_NAMES = [
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
  ];
  const MONTH_NAMES = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
  ];

  const tsDate = new Date(timestamp * 1000);
  const refDate = new Date(reference * 1000);

  // Get UTC date components
  const tsYear = tsDate.getUTCFullYear();
  const tsMonth = tsDate.getUTCMonth();
  const tsDay = tsDate.getUTCDate();
  const tsDow = tsDate.getUTCDay();

  const refYear = refDate.getUTCFullYear();
  const refMonth = refDate.getUTCMonth();
  const refDay = refDate.getUTCDate();

  // Calculate day difference using UTC midnight dates
  const tsMidnight = Date.UTC(tsYear, tsMonth, tsDay);
  const refMidnight = Date.UTC(refYear, refMonth, refDay);
  const dayDiff = Math.round((tsMidnight - refMidnight) / 86400000);

  // Today
  if (dayDiff === 0) {
    return "Today";
  }

  // Yesterday
  if (dayDiff === -1) {
    return "Yesterday";
  }

  // Tomorrow
  if (dayDiff === 1) {
    return "Tomorrow";
  }

  // 2–6 days in the past: "Last <DayName>"
  if (dayDiff >= -6 && dayDiff < -1) {
    return `Last ${DAY_NAMES[tsDow]}`;
  }

  // 2–6 days in the future: "This <DayName>"
  if (dayDiff > 1 && dayDiff <= 6) {
    return `This ${DAY_NAMES[tsDow]}`;
  }

  // Same year: "Month Day"
  if (tsYear === refYear) {
    return `${MONTH_NAMES[tsMonth]} ${tsDay}`;
  }

  // Different year: "Month Day, Year"
  return `${MONTH_NAMES[tsMonth]} ${tsDay}, ${tsYear}`;
}

File: research/experiments/sweep/runs/whenwords-full-kotlin-sonnet45/src/main/kotlin/whenwords/HumanDate.kt
package whenwords

import java.time.Instant
import java.time.ZoneOffset
import java.time.ZonedDateTime
import kotlin.math.round

/**
 * Returns a contextual date string based on proximity to a reference date.
 *
 * @param timestamp Unix epoch seconds (the date to describe)
 * @param reference Unix epoch seconds (the "now" reference point)
 * @return A contextual date label string
 */
fun humanDate(timestamp: Long, reference: Long): String {
    val tsDate = ZonedDateTime.ofInstant(Instant.ofEpochSecond(timestamp), ZoneOffset.UTC)
    val refDate = ZonedDateTime.ofInstant(Instant.ofEpochSecond(reference), ZoneOffset.UTC)

    // Extract components
    val tsYear = tsDate.year
    val tsMonth = tsDate.monthValue
    val tsDay = tsDate.dayOfMonth
    val tsDayOfWeek = tsDate.dayOfWeek.value % 7 // Convert to 0=Sunday, 6=Saturday

    val refYear = refDate.year
    val refMonth = refDate.monthValue
    val refDay = refDate.dayOfMonth

    // Calculate day difference (UTC midnight comparison)
    val tsMidnight = ZonedDateTime.of(tsYear, tsMonth, tsDay, 0, 0, 0, 0, ZoneOffset.UTC)
    val refMidnight = ZonedDateTime.of(refYear, refMonth, refDay, 0, 0, 0, 0, ZoneOffset.UTC)

    val dayDiff = round((tsMidnight.toEpochSecond() - refMidnight.toEpochSecond()).toDouble() / 86400).toInt()

    // Apply rules
    return when (dayDiff) {
        0 -> "Today"
        -1 -> "Yesterday"
        1 -> "Tomorrow"
        in -6..-2 -> "Last ${getDayName(tsDayOfWeek)}"
        in 2..6 -> "This ${getDayName(tsDayOfWeek)}"
        else -> {
            val monthName = getMonthName(tsMonth)
            if (tsYear == refYear) {
                "$monthName $tsDay"
            } else {
                "$monthName $tsDay, $tsYear"
            }
        }
    }
}

private fun getDayName(dayOfWeek: Int): String {
    return when (dayOfWeek) {
        0 -> "Sunday"
        1 -> "Monday"
        2 -> "Tuesday"
        3 -> "Wednesday"
        4 -> "Thursday"
        5 -> "Friday"
        6 -> "Saturday"
        else -> ""
    }
}

private fun getMonthName(month: Int): String {
    return when (month) {
        1 -> "January"
        2 -> "February"
        3 -> "March"
        4 -> "April"
        5 -> "May"
        6 -> "June"
        7 -> "July"
        8 -> "August"
        9 -> "September"
        10 -> "October"
        11 -> "November"
        12 -> "December"
        else -> ""
    }
}
